Проскуряков И.В. Сборник задач по линейной алгебре. СПб.: Лань, 2010.

In [5]:
import numpy as np
from datetime import date

In [139]:
all_problems = [i for i in range(1,123)]
solved_Jupyter = [1, 2, 3, 4]
solved_by_hand = []
solved = solved_Jupyter + solved_by_hand
problems_marked = ['\u0336'.join(' ' + str(i)) if i in solved else str(i) for i in all_problems]

In [131]:
print(*problems_marked)

 ̶1  ̶2  ̶3  ̶4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122


In [138]:
print('Решены в Jupyter:')
print(*solved)

Решены в Jupyter:
1 2 3 4


In [138]:
print('Решены в Jupyter:')
print(*solved)

Решены в Jupyter:
1 2 3 4


In [137]:
print(*[i for i in all_problems if i not in solved])

5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122


In [124]:
'\u0336'.join(' 383')

' ̶3̶8̶3'

# Глава 1. Определители
## §1. Определители 2-го и 3-го порядков

In [146]:
solved_Jupyter = [1, 2, 3, 4, 22, 23, 24, 25, 28, 29, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55]
solved_by_hand = []

problems_num = [i for i in range(1, 123)]
solved = solved_Jupyter + solved_by_hand
problems_marked = ['\u0336'.join(' ' + str(i)) if i in solved else str(i) for i in all_problems]
print(*problems_marked)
print('\nРешены в Jupyter:')
print(*solved_Jupyter)
print('\nРешены в тетради:')
print(*solved_by_hand)
print('\nОсталось решить:')
print(*[i for i in all_problems if i not in solved])

 ̶1  ̶2  ̶3  ̶4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21  ̶2̶2  ̶2̶3  ̶2̶4  ̶2̶5 26 27  ̶2̶8  ̶2̶9 30 31 32 33 34 35 36 37 38 39 40 41 42  ̶4̶3  ̶4̶4  ̶4̶5  ̶4̶6  ̶4̶7  ̶4̶8  ̶4̶9  ̶5̶0  ̶5̶1  ̶5̶2  ̶5̶3  ̶5̶4  ̶5̶5 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122

Решены в Jupyter:
1 2 3 4 22 23 24 25 28 29 43 44 45 46 47 48 49 50 51 52 53 54 55

Решены в тетради:


Осталось решить:
5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 26 27 30 31 32 33 34 35 36 37 38 39 40 41 42 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122


In [9]:
print(f'Сегодня {date.today().isoformat()}')

Сегодня 2021-06-08


Вычислить определители:

In [60]:
problems = {
    1: [
        [5, 2],
        [7, 3]
    ],
    2: [
        [1, 2],
        [3, 4]
    ],
    3: [
        [3, 2],
        [8, 5]
    ],
    4: [
        [6, 9],
        [8, 12]
    ]
}

In [100]:
def det(m):
    """Считает определитель матрицы с целыми 
    элементами"""
    m = np.array(m)
    return round(np.linalg.det(m))

In [104]:
def print_answer1_4(N, m):
    """Выводит ответ на задания 1 - 4"""
    print(f'{N}. Ответ: {int(det(m))}')

In [64]:
for N, m in problems.items():
    print_answer1_4(N, m)

1. Ответ: 1
2. Ответ: -2
3. Ответ: -1
4. Ответ: 0


Задания 5 - 21 в решаются в символьном виде.

Пользуясь определителями, решить системы уравнений:

In [73]:
problems = {
    22: ([
        [2, 5],
        [3, 7]
    ], [1, 2]),
    23: ([
        [2, -3],
        [4, -5]
    ], [4, 10]),
    24: ([
        [5, 7],
        [1, 2]
    ], [1, 0]),
    25: ([
        [4, 7],
        [5, 8]
    ], [-13, -14])
}

In [75]:
def det_solve(m, b):
    """Решение системы c 2 неизвестными по методу Крамера"""
    m = np.array(m)
    det_1 = det(np.column_stack([b, m[:, 1]]))
    det_2 = det(np.column_stack([m[:, 0], b]))
        
    return (det_1 / det(m), det_2 / det(m))

In [105]:
def print_answer22_25(N, m, b):
    """Выводит ответ на задания 22 - 25"""
    print(f'{N}. Ответ: {det_solve(m, b)}')

In [76]:
for N, (m, b) in problems.items():
    print_answer22_25(N, m, b)

22. Ответ: (3.0, -1.0)
23. Ответ: (5.0, 2.0)
24. Ответ: (0.6666666666666666, -0.3333333333333333)
25. Ответ: (2.0, -3.0)


24 задача: опечатка либо в задаче, либо в ответе 

Задания 26 - 27 решаются в символьном виде

Исследовать, будет ли система уравнений опреленна, неопределенна или противоречива:

In [77]:
problems = {
    28: ([
        [4, 6],
        [6, 9]
    ], [2, 3])
}

In [90]:
for N, (m, b) in problems.items():
    D = det(m)
    m = np.array(m)
    det_1 = det(np.column_stack([b, m[:, 1]]))
    det_2 = det(np.column_stack([m[:, 0], b]))
    print(f'{N}. Определители D = {D}, D1 = {det_1}, D2 = {det_2}')
    print('Ответ: Все определители равны нулю, правило Крамера не подходит.'\
          + '\nСистема неопределенна (имеет бесконечно много решений).')

28. Определители D = -0.0, D1 = -0.0, D2 = -0.0
Ответ: Все определители равны нулю, правило Крамера не подходит.
Система неопределенна (имеет бесконечно много решений).


In [91]:
problems = {
    29: ([
        [3, -2],
        [6, -4]
    ], [2, 3])
}

In [95]:
for N, (m, b) in problems.items():
    D = det(m)
    m = np.array(m)
    det_1 = det(np.column_stack([b, m[:, 1]]))
    det_2 = det(np.column_stack([m[:, 0], b]))
    print(f'{N}. Определители D = {D}, D1 = {det_1}, D2 = {det_2}')
    print('Ответ: определитель системы равен нулю, а другие определители нулю не равны.'\
          + '\nСистема противоречива (не имеет решений).')

29. Определители D = 0.0, D1 = -2.0, D2 = -3.0
Ответ: определитель системы равен нулю, а другие определители нулю не равны.
Система противоречива (не имеет решений).


Задания 29 - 37 решаются в символьном виде. Задания 38 - 42 на доказательства. 

Вычислить определители 3-го порядка:

In [106]:
problems = {
    43: [
        [2, 1, 3],
        [5, 3, 2],
        [1, 4, 3]
    ],
    44: [
        [3, 2, 1],
        [2, 5, 3],
        [3, 4, 2]
    ],
    45: [
        [4, -3, 5],
        [3, -2, 8],
        [1, -7, -5]
    ],
    46: [
        [3, 2, -4],
        [4, 1, -2],
        [5, 2, -3]
    ],
    47: [
        [3, 4, -5],
        [8, 7, -2],
        [2, -1, 8]
    ],
    48: [
        [4, 2, -1],
        [5, 3, -2],
        [3, 2, -1]
    ],
    49: [
        [1, 1, 1],
        [1, 2, 3],
        [1, 3, 6]
    ],
    50: [
        [0, 1, 1],
        [1, 0, 1],
        [1, 1, 0]
    ],
    51: [
        [5, 6, 3],
        [0, 1, 0],
        [7, 4, 5]
    ],
    52: [
        [2, 0, 3],
        [7, 1, 6],
        [6, 0, 5]
    ],
    53: [
        [1, 5, 25],
        [1, 7, 49],
        [1, 8, 64]
    ],
    54: [
        [1, 1, 1],
        [4, 5, 9],
        [16, 25, 81]
    ],
    55: [
        [1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]
    ]
}

In [109]:
def print_answer43_55(N, m):
    """Выводит ответ на задания 43 - 55"""
    print(f'{N}. Ответ: {int(det(m))}.')

In [110]:
for N, m in problems.items():
    print_answer43_55(N, m)

43. Ответ: 40.
44. Ответ: -3.
45. Ответ: 100.
46. Ответ: -5.
47. Ответ: 0.
48. Ответ: 1.
49. Ответ: 1.
50. Ответ: 2.
51. Ответ: 4.
52. Ответ: -8.
53. Ответ: 6.
54. Ответ: 20.
55. Ответ: 0.


Задания 56 - 73 решаются в символьном виде.

Пользуясь определителями, решить системы уравнений: 74 - 77

Задания 78 - 81 решаются в символьном виде. 

Исследовать, будет ли система уравнений опреленна, неопределенна или противоречива: 82 - 85

Задания 90 - 122 решаются в символьном виде. 

## §4. Вычисление определителей с числовыми элементами

In [147]:
solved_Jupyter = []
solved_by_hand = []

problems_num = [i for i in range(257, 279)]
solved = solved_Jupyter + solved_by_hand
problems_marked = ['\u0336'.join(' ' + str(i)) if i in solved else str(i) for i in all_problems]
print(*problems_marked)
print('\nРешены в Jupyter:')
print(*solved_Jupyter)
print('\nРешены в тетради:')
print(*solved_by_hand)
print('\nОсталось решить:')
print(*[i for i in all_problems if i not in solved])

1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122

Решены в Jupyter:


Решены в тетради:


Осталось решить:
1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122


Вычислить определители: 257 - 278